backend/src/DataSources/posts-data-source.js

```diff
@@ -0,0 +1,87 @@
+const {DataSource} = require('apollo-datasource');
+const {UserInputError} = require('apollo-server');
+const crypto = require('crypto');
+
+class Post {
+  constructor (data) {
+    this.id = crypto.randomBytes(16).toString('hex');
+    this.voters = [];
+    this.votes = 0;
+    Object.assign(this, data);
+  }
+}
+
+class PostsDataSource extends DataSource {
+  
+  constructor() {
+    super();
+    this.posts = [];
+  }
+
+  async allPosts () {
+    return this.posts;
+  }
+
+  async createPost (data, currUser) {
+
+    if(currUser){
+      const newPost = new Post({...data, author:currUser});
+      this.posts.push(newPost);
+      
+      return newPost;
+    }
+
+    return null;
```

Comment:
```suggestion
    throw AuthenticationError("You must be logged in");
```
https://www.apollographql.com/docs/apollo-server/data/errors/

If you implement a permission middleware around your resolvers, you won't even need this check here.